flang/lib/Optimizer/CodeGen/PreCGRewrite.cpp

```diff
@@ -276,7 +276,29 @@ class DeclareOpConversion : public mlir::OpRewritePattern<fir::DeclareOp> {
   mlir::LogicalResult
   matchAndRewrite(fir::DeclareOp declareOp,
                   mlir::PatternRewriter &rewriter) const override {
-    rewriter.replaceOp(declareOp, declareOp.getMemref());
+    auto loc = declareOp.getLoc();
+    llvm::SmallVector<mlir::Value> shapeOpers;
+    llvm::SmallVector<mlir::Value> shiftOpers;
+    if (auto shapeVal = declareOp.getShape()) {
+      if (auto shapeOp = mlir::dyn_cast<fir::ShapeOp>(shapeVal.getDefiningOp()))
+        populateShape(shapeOpers, shapeOp);
+      else if (auto shiftOp =
+                   mlir::dyn_cast<fir::ShapeShiftOp>(shapeVal.getDefiningOp()))
+        populateShapeAndShift(shapeOpers, shiftOpers, shiftOp);
+      else if (auto shiftOp =
+                   mlir::dyn_cast<fir::ShiftOp>(shapeVal.getDefiningOp()))
+        populateShift(shiftOpers, shiftOp);
+      else
+        return mlir::failure();
+    }
+    // FIXME: Add FortranAttrs and CudaAttrs
+    auto xDeclOp = rewriter.create<fir::cg::XDeclareOp>(
+        loc, declareOp.getType(), declareOp.getMemref(), shapeOpers, shiftOpers,
```

Comment:
Maybe this should only be created if there is a DILocalVariableAttr fused with the location of the DeclareOp.

Creating MLIR operations, especially with many arguments, has a compile time cost (both memory and time), so I would be inclined to start by only doing it when we know it is needed, which is easy here currently. That way, there is no extra compile time cost for filed compiled without -g.

Future use cases can enable this translation when they arise.

Reply:
This pre-cg-rewrite runs before `AddDebugInfo` so we would not have created `FusedLoc` by that time. I am thinking that we could run a cleanup pass before pre-cg-rewrite that removes the `DeclareOp` if debug info is disabled. Does that sound like the right approach to you?

Reply:
Good point. I think adding a pass option "preserveDeclare" to pre-cg-rewrite  passwould be simpler and cheaper then. It can be set in createDefaultFIRCodeGenPassPipeline [here](https://github.com/llvm/llvm-project/blob/37f3661f4e8d180dc702ab42fa015bc3c21ceb51/flang/include/flang/Tools/CLOptions.inc#L308) given the debug level.

Reply:
Thanks for the suggestions. I have added the option in 2c566dd. Also updated the tests to check with this option turned off and on.